notebooks/knn.ipynb

```diff
@@ -0,0 +1,399 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "61d09a34-15a1-456a-a6dd-fc9abad2716a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "item_id_df, query_id_df, neighbor_df = knn_model.kneighbors(query_df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fde68c05-5cde-4b1e-8ed3-ae5fe2d7015b",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# original item_df is returned with unique identifiers\n",
+    "item_id_df.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "281c204b-5804-4ae8-9064-e6fe08355618",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# original query_df is returned with unique identifiers\n",
+    "query_id_df.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "457b702d-1af0-4c38-b1b5-dd1a1f5f1dec",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# neighbor_df shows the nearest item vectors for each query vector, represented by their unique ids and distances.\n",
+    "neighbor_df.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "62f81b49-6ff3-4656-8b0c-27b32a3f60de",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "# change the value of 'k'\n",
+    "knn_model.setK(3)\n",
+    "_, _, neighbor_df = knn_model.kneighbors(query_df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a52e7e03-0bac-4378-b219-eda5d00523da",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "neighbor_df.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "21d0f88c-8ba0-4f25-a42f-45c864972f54",
+   "metadata": {},
+   "source": [
+    "#### exactNearestNeighborsJoin\n",
+    "\n",
+    "This API returns a join of the query vectors and their `k` closest item vectors."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "6dc0fd29-f25e-4d89-95fe-ebd2efbfe9ea",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "result_df = knn_model.exactNearestNeighborsJoin(query_df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5196d375-db08-4292-a865-61fecd07fe41",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "result_df.orderBy(\"query_df\", \"item_df\").show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8cd56670-7633-4fe6-ab75-0fd680c63baa",
+   "metadata": {},
+   "source": [
+    "# PySpark\n",
+    "\n",
+    "Note: PySpark does not have an exact kNN implementation, but it does have an LSH-based approximate nearest neighbors implementation, shown here."
```

Comment:
Maybe add a clarifying comment about why this is here.  Like to illustrate similarity of apis.   Ideally the results could be compared but that would require a more elaborate tuning of parameters here.

Reply:
Done.